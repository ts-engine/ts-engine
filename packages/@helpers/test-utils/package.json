{
  "name": "@helpers/test-utils",
  "version": "1.1.1",
  "license": "MIT",
  "private": true,
  "main": "dist/main.cjs.js",
  "module": "dist/main.esm.js",
  "types": "dist/main.d.ts",
  "scripts": {
    "build": "rm -rf dist && ts-engine build --library",
    "lint": "ts-engine lint",
    "typecheck": "ts-engine typecheck"
  },
  "devDependencies": {
    "@ts-engine/cli": "1.1.1"
  },
  "dependencies": {
    "chalk": "^4.0.0",
    "fs-extra": "^9.0.0",
    "strip-ansi": "^6.0.0",
    "terminate": "^2.1.2"
  }
}
