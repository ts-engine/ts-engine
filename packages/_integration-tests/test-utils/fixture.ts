import path from "path";
import { spawnSync } from "child_process";
import fs from "fs-extra";
import { parseArgsStringToArgv } from "string-argv";
import stripAnsi from "strip-ansi";

interface RunCliResult {
  status: number;
  stdout: string;
  stderr: string;
}

interface RunToolOptions {
  name: string;
  cwd: string;
  args: string;
}

const runTool = (options: RunToolOptions) => {
  const result = spawnSync(options.name, parseArgsStringToArgv(options.args), {
    cwd: options.cwd,
    encoding: "utf8",
  });

  return {
    status: result.status,
    stdout: stripAnsi(result.stdout),
    stderr: stripAnsi(result.stderr),
  };
};

interface Fixture {
  reset: () => Promise<void>;
  runTse: (args: string) => RunCliResult;
  runNode: (args: string) => RunCliResult;
  writeFile: (filepath: string, content: string) => Promise<void>;
  removeFile: (filepath: string) => Promise<void>;
  readFile: (filepath: string) => Promise<string>;
}

type FixtureName =
  | "build-normal"
  | "build-minify"
  | "build-react"
  | "build-syntax-error"
  | "build-type-error"
  | "lint-error"
  | "lint-fix"
  | "lint-pass"
  | "lint-warning"
  | "lint-no-files"
  | "run-normal"
  | "run-minify"
  | "run-react"
  | "run-syntax-error"
  | "run-type-error"
  | "test-error"
  | "test-pass"
  | "test-config"
  | "test-setup-ts"
  | "test-setup-js"
  | "test-no-tests"
  | "version";

export const createFixture = (name: FixtureName): Fixture => {
  const dir = path.resolve("../_integration-fixtures/", name);
  const distDir = path.resolve(dir, "dist");

  const reset = async () => {
    await fs.remove(distDir);
  };

  const runTse = (args: string) => {
    return runTool({
      name: "yarn",
      cwd: dir,
      args: `run tse ${args}`,
    });
  };

  const runNode = (args: string) => {
    return runTool({
      name: "node",
      cwd: dir,
      args,
    });
  };

  const writeFile = async (filepath: string, content: string) => {
    await fs.writeFile(path.resolve(dir, filepath), content, {
      encoding: "utf8",
    });
  };

  const removeFile = async (filepath: string) => {
    await fs.remove(path.resolve(dir, filepath));
  };

  const readFile = async (filepath: string) => {
    return await fs.readFile(path.resolve(dir, filepath), { encoding: "utf8" });
  };

  return {
    reset,
    runTse,
    runNode,
    writeFile,
    removeFile,
    readFile,
  };
};
