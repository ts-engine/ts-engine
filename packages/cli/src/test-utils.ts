import { parseArgsStringToArgv } from "string-argv";
import { cli } from "./cli";

export const testConsole = {
  log: console.debug,
};

interface RunCliResult {
  stdout: string[];
  stderr: string[];
  exitCode: number;
}

export const runCli = async (args: string): Promise<RunCliResult> => {
  // capture stdout/stderr console logs
  const stderr: string[] = [];
  const stdout: string[] = [];

  const stderrSpy = jest.fn((message: any) => {
    if (message) {
      stderr.push(message.toString());
    }
  });

  const stdoutSpy = jest.fn((message: any) => {
    if (message) {
      stdout.push(message.toString());
    }
  });

  jest.spyOn(console, "error").mockImplementation(stderrSpy);
  jest.spyOn(console, "info").mockImplementation(stdoutSpy);
  jest.spyOn(console, "log").mockImplementation(stdoutSpy);
  jest.spyOn(console, "warn").mockImplementation(stdoutSpy);

  // capture exit code
  let exitCode = 0;
  const exitPromise = new Promise((resolve) => {
    // @ts-ignore
    jest.spyOn(process, "exit").mockImplementation((code: number) => {
      // only assign if the code is still 0 (falsy) as
      // process.exit can only be called once
      exitCode = exitCode || code;
      resolve();
    });
  });

  // run cli
  cli(parseArgsStringToArgv(args));

  // let async code complete
  await exitPromise;

  // reset mocks
  (console.error as jest.Mock).mockRestore();
  (console.info as jest.Mock).mockRestore();
  (console.log as jest.Mock).mockRestore();
  (console.warn as jest.Mock).mockRestore();

  // @ts-ignore
  (process.exit as jest.Mock).mockRestore();

  // collate results
  return {
    stderr,
    stdout,
    exitCode,
  };
};

export const matchLog = (regex: RegExp | string, logs: string[]) => {
  return logs.find((log) => log.match(regex));
};
