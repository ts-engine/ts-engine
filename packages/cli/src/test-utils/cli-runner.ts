import { cli } from "../cli";

interface RunCliResult {
  stdout: string[];
  stderr: string[];
  exitCode: number;
}

export const runCli = async (...args: string[]): Promise<RunCliResult> => {
  // capture stdout/stderr console logs
  const stderr: string[] = [];
  const stdout: string[] = [];

  const stderrSpy = jest.fn((message: any) => {
    stderr.push(message);
  });

  const stdoutSpy = jest.fn((message: any) => {
    stdout.push(message);
  });

  console.error = jest.fn(stderrSpy);
  console.info = jest.fn(stdoutSpy);
  console.log = jest.fn(stdoutSpy);
  console.warn = jest.fn(stdoutSpy);

  // capture exit code
  let exitCode = 0;

  // @ts-ignore
  process.exit = jest.fn((code: number) => (exitCode = code));

  // run cli
  cli(args);

  // let async code complete
  await new Promise((resolve) => setImmediate(resolve));

  // reset mocks
  (console.error as jest.Mock).mockRestore();
  (console.info as jest.Mock).mockRestore();
  (console.log as jest.Mock).mockRestore();
  (console.warn as jest.Mock).mockRestore();

  // @ts-ignore
  (process.exit as jest.Mock).mockRestore();

  // collate results
  return {
    stderr,
    stdout,
    exitCode,
  };
};
