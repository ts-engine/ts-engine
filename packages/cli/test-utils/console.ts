interface MockConsoleResult {
  log: jest.Mock;
  error: jest.Mock;
  warn: jest.Mock;
  reset: () => void;
}

export const mockConsole = (): MockConsoleResult => {
  const consoleLogMock = jest.fn();
  const consoleErrorMock = jest.fn();
  const consoleWarnMock = jest.fn();

  console.log = consoleLogMock;
  console.error = consoleErrorMock;
  console.warn = consoleWarnMock;

  return {
    log: consoleLogMock,
    error: consoleErrorMock,
    warn: consoleWarnMock,
    reset: () => {
      consoleLogMock.mockReset();
      consoleErrorMock.mockReset();
      consoleWarnMock.mockReset();
    },
  };
};
