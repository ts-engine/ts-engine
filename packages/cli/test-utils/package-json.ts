export const createPackageJsonMock = () => {
  return {
    name: "example",
    version: "1.0.0",
    description: "Example package.json",
  };
};
