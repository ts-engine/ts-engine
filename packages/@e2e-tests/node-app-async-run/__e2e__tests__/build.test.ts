import { fileSystem, runCliCommand } from "@helpers/test-utils";

describe("build", () => {
  beforeEach(async () => {
    await fileSystem.deleteDir("dist");
  });

  it("should build the code", async () => {
    const runner = runCliCommand("yarn run ts-engine build --node-app");

    // Wait for tool to complete
    const statusCode = await runner.waitForStatusCode();

    // Should exit successfully
    expect(statusCode).toBe(0);

    // Built file is written to file system
    expect(await fileSystem.fileExists("dist/main.js")).toBe(true);
  });

  it("built app should work", async () => {
    const runner = runCliCommand("yarn run ts-engine build --node-app");

    // Wait for tool to complete
    await runner.waitForStatusCode();

    // Run the app
    const appRunner = runCliCommand("node dist/main.js");

    // Wait for app to complete
    const statusCode = await appRunner.waitForStatusCode();

    // Should exist successfully
    expect(statusCode).toBe(0);

    // Should have printed message
    expect(appRunner.stdoutLines).toContain("Hello Lee!");
  }, 10000);
});
