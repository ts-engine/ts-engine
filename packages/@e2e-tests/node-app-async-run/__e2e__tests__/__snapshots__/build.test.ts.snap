// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build should build the code 1`] = `
"\\"use strict\\";function t(t,r){return t(r={exports:{}},r.exports),r.exports}var r=t((function(t){function r(e){return\\"function\\"==typeof Symbol&&\\"symbol\\"==typeof Symbol.iterator?t.exports=r=function(t){return typeof t}:t.exports=r=function(t){return t&&\\"function\\"==typeof Symbol&&t.constructor===Symbol&&t!==Symbol.prototype?\\"symbol\\":typeof t},r(e)}t.exports=r})),e=t((function(t){var e=function(t){var e=Object.prototype,n=e.hasOwnProperty,o=\\"function\\"==typeof Symbol?Symbol:{},i=o.iterator||\\"@@iterator\\",a=o.asyncIterator||\\"@@asyncIterator\\",c=o.toStringTag||\\"@@toStringTag\\";function u(t,r,e,n){var o=r&&r.prototype instanceof h?r:h,i=Object.create(o.prototype),a=new E(n||[]);return i._invoke=function(t,r,e){var n=\\"suspendedStart\\";return function(o,i){if(\\"executing\\"===n)throw new Error(\\"Generator is already running\\");if(\\"completed\\"===n){if(\\"throw\\"===o)throw i;return S()}for(e.method=o,e.arg=i;;){var a=e.delegate;if(a){var c=x(a,e);if(c){if(c===s)continue;return c}}if(\\"next\\"===e.method)e.sent=e._sent=e.arg;else if(\\"throw\\"===e.method){if(\\"suspendedStart\\"===n)throw n=\\"completed\\",e.arg;e.dispatchException(e.arg)}else\\"return\\"===e.method&&e.abrupt(\\"return\\",e.arg);n=\\"executing\\";var u=f(t,r,e);if(\\"normal\\"===u.type){if(n=e.done?\\"completed\\":\\"suspendedYield\\",u.arg===s)continue;return{value:u.arg,done:e.done}}\\"throw\\"===u.type&&(n=\\"completed\\",e.method=\\"throw\\",e.arg=u.arg)}}}(t,e,a),i}function f(t,r,e){try{return{type:\\"normal\\",arg:t.call(r,e)}}catch(t){return{type:\\"throw\\",arg:t}}}t.wrap=u;var s={};function h(){}function l(){}function p(){}var y={};y[i]=function(){return this};var v=Object.getPrototypeOf,d=v&&v(v(_([])));d&&d!==e&&n.call(d,i)&&(y=d);var m=p.prototype=h.prototype=Object.create(y);function g(t){[\\"next\\",\\"throw\\",\\"return\\"].forEach((function(r){t[r]=function(t){return this._invoke(r,t)}}))}function w(t,e){var o;this._invoke=function(i,a){function c(){return new e((function(o,c){!function o(i,a,c,u){var s=f(t[i],t,a);if(\\"throw\\"!==s.type){var h=s.arg,l=h.value;return l&&\\"object\\"===r(l)&&n.call(l,\\"__await\\")?e.resolve(l.__await).then((function(t){o(\\"next\\",t,c,u)}),(function(t){o(\\"throw\\",t,c,u)})):e.resolve(l).then((function(t){h.value=t,c(h)}),(function(t){return o(\\"throw\\",t,c,u)}))}u(s.arg)}(i,a,o,c)}))}return o=o?o.then(c,c):c()}}function x(t,r){var e=t.iterator[r.method];if(void 0===e){if(r.delegate=null,\\"throw\\"===r.method){if(t.iterator.return&&(r.method=\\"return\\",r.arg=void 0,x(t,r),\\"throw\\"===r.method))return s;r.method=\\"throw\\",r.arg=new TypeError(\\"The iterator does not provide a 'throw' method\\")}return s}var n=f(e,t.iterator,r.arg);if(\\"throw\\"===n.type)return r.method=\\"throw\\",r.arg=n.arg,r.delegate=null,s;var o=n.arg;return o?o.done?(r[t.resultName]=o.value,r.next=t.nextLoc,\\"return\\"!==r.method&&(r.method=\\"next\\",r.arg=void 0),r.delegate=null,s):o:(r.method=\\"throw\\",r.arg=new TypeError(\\"iterator result is not an object\\"),r.delegate=null,s)}function L(t){var r={tryLoc:t[0]};1 in t&&(r.catchLoc=t[1]),2 in t&&(r.finallyLoc=t[2],r.afterLoc=t[3]),this.tryEntries.push(r)}function b(t){var r=t.completion||{};r.type=\\"normal\\",delete r.arg,t.completion=r}function E(t){this.tryEntries=[{tryLoc:\\"root\\"}],t.forEach(L,this),this.reset(!0)}function _(t){if(t){var r=t[i];if(r)return r.call(t);if(\\"function\\"==typeof t.next)return t;if(!isNaN(t.length)){var e=-1,o=function r(){for(;++e<t.length;)if(n.call(t,e))return r.value=t[e],r.done=!1,r;return r.value=void 0,r.done=!0,r};return o.next=o}}return{next:S}}function S(){return{value:void 0,done:!0}}return l.prototype=m.constructor=p,p.constructor=l,p[c]=l.displayName=\\"GeneratorFunction\\",t.isGeneratorFunction=function(t){var r=\\"function\\"==typeof t&&t.constructor;return!!r&&(r===l||\\"GeneratorFunction\\"===(r.displayName||r.name))},t.mark=function(t){return Object.setPrototypeOf?Object.setPrototypeOf(t,p):(t.__proto__=p,c in t||(t[c]=\\"GeneratorFunction\\")),t.prototype=Object.create(m),t},t.awrap=function(t){return{__await:t}},g(w.prototype),w.prototype[a]=function(){return this},t.AsyncIterator=w,t.async=function(r,e,n,o,i){void 0===i&&(i=Promise);var a=new w(u(r,e,n,o),i);return t.isGeneratorFunction(e)?a:a.next().then((function(t){return t.done?t.value:a.next()}))},g(m),m[c]=\\"Generator\\",m[i]=function(){return this},m.toString=function(){return\\"[object Generator]\\"},t.keys=function(t){var r=[];for(var e in t)r.push(e);return r.reverse(),function e(){for(;r.length;){var n=r.pop();if(n in t)return e.value=n,e.done=!1,e}return e.done=!0,e}},t.values=_,E.prototype={constructor:E,reset:function(t){if(this.prev=0,this.next=0,this.sent=this._sent=void 0,this.done=!1,this.delegate=null,this.method=\\"next\\",this.arg=void 0,this.tryEntries.forEach(b),!t)for(var r in this)\\"t\\"===r.charAt(0)&&n.call(this,r)&&!isNaN(+r.slice(1))&&(this[r]=void 0)},stop:function(){this.done=!0;var t=this.tryEntries[0].completion;if(\\"throw\\"===t.type)throw t.arg;return this.rval},dispatchException:function(t){if(this.done)throw t;var r=this;function e(e,n){return a.type=\\"throw\\",a.arg=t,r.next=e,n&&(r.method=\\"next\\",r.arg=void 0),!!n}for(var o=this.tryEntries.length-1;o>=0;--o){var i=this.tryEntries[o],a=i.completion;if(\\"root\\"===i.tryLoc)return e(\\"end\\");if(i.tryLoc<=this.prev){var c=n.call(i,\\"catchLoc\\"),u=n.call(i,\\"finallyLoc\\");if(c&&u){if(this.prev<i.catchLoc)return e(i.catchLoc,!0);if(this.prev<i.finallyLoc)return e(i.finallyLoc)}else if(c){if(this.prev<i.catchLoc)return e(i.catchLoc,!0)}else{if(!u)throw new Error(\\"try statement without catch or finally\\");if(this.prev<i.finallyLoc)return e(i.finallyLoc)}}}},abrupt:function(t,r){for(var e=this.tryEntries.length-1;e>=0;--e){var o=this.tryEntries[e];if(o.tryLoc<=this.prev&&n.call(o,\\"finallyLoc\\")&&this.prev<o.finallyLoc){var i=o;break}}i&&(\\"break\\"===t||\\"continue\\"===t)&&i.tryLoc<=r&&r<=i.finallyLoc&&(i=null);var a=i?i.completion:{};return a.type=t,a.arg=r,i?(this.method=\\"next\\",this.next=i.finallyLoc,s):this.complete(a)},complete:function(t,r){if(\\"throw\\"===t.type)throw t.arg;return\\"break\\"===t.type||\\"continue\\"===t.type?this.next=t.arg:\\"return\\"===t.type?(this.rval=this.arg=t.arg,this.method=\\"return\\",this.next=\\"end\\"):\\"normal\\"===t.type&&r&&(this.next=r),s},finish:function(t){for(var r=this.tryEntries.length-1;r>=0;--r){var e=this.tryEntries[r];if(e.finallyLoc===t)return this.complete(e.completion,e.afterLoc),b(e),s}},catch:function(t){for(var r=this.tryEntries.length-1;r>=0;--r){var e=this.tryEntries[r];if(e.tryLoc===t){var n=e.completion;if(\\"throw\\"===n.type){var o=n.arg;b(e)}return o}}throw new Error(\\"illegal catch attempt\\")},delegateYield:function(t,r,e){return this.delegate={iterator:_(t),resultName:r,nextLoc:e},\\"next\\"===this.method&&(this.arg=void 0),s}},t}(t.exports);try{regeneratorRuntime=e}catch(t){Function(\\"r\\",\\"regeneratorRuntime = r\\")(e)}}));function n(t,r,e,n,o,i,a){try{var c=t[i](a),u=c.value}catch(t){return void e(t)}c.done?r(u):Promise.resolve(u).then(n,o)}var o=function(t){return function(){var r=this,e=arguments;return new Promise((function(o,i){var a=t.apply(r,e);function c(t){n(a,o,i,c,u,\\"next\\",t)}function u(t){n(a,o,i,c,u,\\"throw\\",t)}c(void 0)}))}};(function(){var t=o(e.mark((function t(){return e.wrap((function(t){for(;;)switch(t.prev=t.next){case 0:return t.t0=console,t.next=3,r=\\"Lee\\",Promise.resolve(\\"Hello \\".concat(r,\\"!\\"));case 3:t.t1=t.sent,t.t0.log.call(t.t0,t.t1);case 5:case\\"end\\":return t.stop()}var r}),t)})));return function(){return t.apply(this,arguments)}})()();
"
`;
